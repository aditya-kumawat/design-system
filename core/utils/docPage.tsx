// @ts-nocheck

import * as React from 'react';
import { Title, Subtitle, Heading, Subheading, Props, Description, Preview, Source, PropsProps } from '@storybook/addon-docs/blocks';
import { renderToStaticMarkup } from 'react-dom/server';
import reactElementToJSXString from 'react-element-to-jsx-string';
import { html as beautifyHTML } from 'js-beautify';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { docco, dark, a11yDark, githubGist, dracula, vs2015 } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import * as DS from '@/';
import { Button, Card, TabsWrapper, Tab } from '@/';
import vsDark from 'prism-react-renderer/themes/vsDark';

import {
  LiveProvider,
  LiveEditor,
  LiveError,
  LivePreview,
  withLive
} from 'react-live';

export interface Example {
  title: string;
  description?: string;
  imports: string[];
  component: React.ReactNode;
}

interface DocPageProps {
  title?: string;
  description?: React.ReactNode;
  props?: PropsProps;
  customCode?: string;
}

const beautifyHTMLOptions = {
  indent_size: 2,
  wrap_line_length: 0,
  preserve_newlines: true,
  jslint_happy: true,
  end_with_newline: false,
  indent_inner_html: true
};

const beautifyJSXOptions = {
  ...beautifyHTMLOptions,
  e4x: true
};

const JSXtoStringOptions = {
  filterProps: (val: any, key: string) => {
    if (!val) return false;
    if (val && val.name === 'actionHandler') return false;
    // if(val && typeof val === 'function') return false;
    return true;
  },
  showFunctions: true,
  functionValue: _fn => _ => { },
  // maxInlineAttributesLineLength: 10
};

const copyCode = (val: string) => navigator.clipboard.writeText(val);

const importsToStr = (imports: string[]): string => `import { ${imports.join(', ')} } from 'design-system';`;

const CopyComp = props => {
  const { onClick } = props;
  return (
    <div
      style={{
        position: 'absolute',
        right: '0',
        top: '0',
        zIndex: 10
      }}
    >
      <Button size="small" style={{ borderRadius: '0' }} appearance="basic" onClick={onClick}>Copy</Button>
    </div>
  );
};

const renderCodeBlock = (val: string) => (
  <pre style={{ position: 'relative', fontFamily: 'monospace', fontSize: '13px', background: '#f8f8f8' }}>
    <CopyComp onClick={() => copyCode(val)} />
    <SyntaxHighlighter language="javascript" style={vs2015} showLineNumbers={true}>
      {val}
    </SyntaxHighlighter>
  </pre>
);

const StoryComp = props => {
  const {
    customCode
  } = props;

  const { storyId } = __STORYBOOK_STORY_STORE__.getSelection();
  const story = __STORYBOOK_STORY_STORE__.fromId(storyId);
  const comp = story.storyFn().props.children.props.children.props.children;
  const sp = story.parameters;
  const imports = [sp.component, ...(sp.subcomponents ? Object.values(sp.subcomponents) : [])];

  const importString = `// ${importsToStr(imports.map(i => i.displayName))}`;
  const jsx = customCode ? customCode : `${beautifyHTML(reactElementToJSXString(comp, JSXtoStringOptions), beautifyJSXOptions)}`;
  const html = beautifyHTML(renderToStaticMarkup(comp), beautifyHTMLOptions);

  const [activeTab, setActiveTab] = React.useState<number>(0);
  const [jsxCode, setJsxCode] = React.useState<string>(`${importString}\n\n${jsx}`);
  const [htmlCode, setHtmlCode] = React.useState<string>(`${html}`);

  const importScope = imports.reduce((out, curr) => {
    out[curr.displayName] = curr;
    return out;
  }, {});

  const TabsWrap = withLive(({ live }) => {
    const {
      error,
      element: Element,
    } = live;

    React.useEffect(() => {
      if (!error) {
        const editor = document.querySelector('.npm__react-simple-code-editor__textarea');

        if (editor) {
          const jsxValue = editor.value;
          setJsxCode(jsxValue);

          try {
            const htmlValue = beautifyHTML(renderToStaticMarkup(<Element />), beautifyHTMLOptions);
            setHtmlCode(htmlValue);
          } catch (e) { }
        }
      }
    }, [live]);

    return (
      <></>
    );
  });

  return (
    <>
      <Card
        shadow="light"
      >
        <LiveProvider code={jsxCode} scope={{ ...DS, ...importScope }}>
          <Preview
            style={{
              marginTop: '0px',
              marginBottom: '0px'
            }}
            withSource="none"
            withToolbar={true}
          >
            <LivePreview />
          </Preview>
          <TabsWrap />
          <TabsWrapper
            activeTab={activeTab}
            onTabChange={tab => setActiveTab(tab)}
          >
            <Tab label={'React'}>
              <div style={{ position: 'relative' }}>
                <CopyComp
                  onClick={() => {
                    const editor = document.querySelector('.npm__react-simple-code-editor__textarea');
                    if (editor) copyCode(editor.value);
                  }}
                />
                <LiveEditor theme={vsDark} />
                <LiveError />
              </div>
            </Tab>
            <Tab label={'HTML'}>
              {renderCodeBlock(htmlCode)}
            </Tab>
          </TabsWrapper>
        </LiveProvider>
      </Card>
    </>
  );
};

export const docPage = () => {
  const { storyId } = __STORYBOOK_STORY_STORE__.getSelection();
  const story = __STORYBOOK_STORY_STORE__.fromId(storyId);
  const comp = story.storyFn().props.children.props.children.props.children;
  const sp = story.parameters;

  const {
    title,
    description,
    props: propsAttr,
    customCode,
    noStory
  } = sp.docs.docPage;

  const separatorIndex = title?.indexOf('|');

  return (
    <>
      <Title>{title && separatorIndex !== -1 ? title.slice(separatorIndex + 1) : title}</Title>
      <br />

      <Description>
        {description}
      </Description>
      <br />

      {!noStory && (
        <>
          <Heading>Story</Heading>
          <StoryComp key={storyId} customCode={customCode} />
          <br />
        </>
      )}

      <Heading>PropTable</Heading>
      <Props {...propsAttr} />
    </>
  );
};

export default docPage;
