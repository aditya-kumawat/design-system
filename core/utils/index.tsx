import * as positionHelper from './positionHelper';
import PopperWrapper from './PopperWrapper';

export { positionHelper,  PopperWrapper };
export * from './css';
