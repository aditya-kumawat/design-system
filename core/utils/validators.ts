export const dateValidator = (format: string, val: string): boolean => {
  const validate = (date: number, month: number, year: number): boolean => {
    // Check the ranges of month and year
    if (year < 1000 || year > 3000 || month == 0 || month > 12)
      return false;

    var monthLength = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

    // Adjust for leap years
    if (year % 400 == 0 || (year % 100 != 0 && year % 4 == 0))
      monthLength[1] = 29;

    // Check the range of the day
    return date > 0 && date <= monthLength[month - 1];
  }

  switch (format) {
    case 'dd/mm/yy':
      var p = val.split('/');
      var date = +p[0];
      var month = +p[1];
      var year = +p[2];
      return validate(date, month, year);

    case 'mm/dd/yy':
      var p = val.split('/');
      var date = +p[1];
      var month = +p[0];
      var year = +p[2];
      return validate(date, month, year);

    case 'yy/mm/dd':
      var p = val.split('/');
      var date = +p[2];
      var month = +p[1];
      var year = +p[0];
      return validate(date, month, year);

    case 'dd-mm-yy':
      var p = val.split('-');
      var date = +p[0];
      var month = +p[1];
      var year = +p[2];
      return validate(date, month, year);

    case 'mm-dd-yy':
      var p = val.split('-');
      var date = +p[1];
      var month = +p[0];
      var year = +p[2];
      return validate(date, month, year);

    case 'yy-mm-dd':
      var p = val.split('-');
      var date = +p[2];
      var month = +p[1];
      var year = +p[0];
      return validate(date, month, year);

    default:
      return false;
  }
}

const e: Record<string, any> = {
  date: dateValidator
}
export default e;