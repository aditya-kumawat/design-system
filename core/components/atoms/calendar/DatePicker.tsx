import * as React from 'react';
import Calendar from './Calendar';
import { View, Day, DateType, DateFormat } from './types';
import Popover from '@/components/molecules/popover';
import InputMask from '@/components/molecules/inputMask';
import { getDateInfo, convertToDate } from './utility';
import validators from '@/utils/validators';

export interface IDatePickerProps {
  onDateChange?: (date?: Date) => void;
  jumpView?: boolean;
  date?: DateType;
  firstDayOfWeek?: Day;
  view?: View;
  disabledBefore?: DateType;
  disabledAfter?: DateType;
  yearNav?: number;
  monthNav?: number;
  withInput?: boolean;
  inputFormat?: DateFormat;
  outputFormat?: DateFormat;
}

export const DatePicker: React.FunctionComponent<IDatePickerProps> = props => {
  const {
    date: dateProp,
    withInput = false,
    inputFormat = 'mm/dd/yy',
    outputFormat = 'mm/dd/yy',
    onDateChange,
    ...rest
  } = props;

  const [date, setDate] = React.useState<Date | undefined>(dateProp ? convertToDate(dateProp) : undefined);

  if (withInput) {
    const translateToString = (format: string, d: Date): string => {
      const {
        year,
        month,
        date
      } = getDateInfo(d);

      const separator = format.includes('/') ? '/' : '-';
      const val = format.split(separator).reduce((out, curr) => {
        switch (curr) {
          case 'mm':
            out += (month < 10 && '0') + (month + 1);
            break;
          case 'yy':
            out += year;
            break;
          case 'dd':
            out += (date < 10 && '0') + date;
            break;
        }
        return out;
      }, '');

      return val;
    }

    const translateToDate = (format: string, val: string): Date | undefined => {
      if (validators.date(inputFormat, val)) {
        const separator = format.includes('/') ? '/' : '-';

        let year: number = -1,
          month: number = -1,
          date: number = -1;
        const v = val.split(separator);
        format.split(separator).forEach((f, i) => {
          switch (f) {
            case 'mm':
              month = +v[i] - 1;
              break;
            case 'yy':
              year = +v[i];
              break;
            case 'dd':
              date = +v[i];
              break;
          }
        });
        const d = convertToDate({ year, month, date });
        return d;
      } else {
        return undefined;
      }
    }

    const onChangeHandler = (_e: React.ChangeEvent<HTMLInputElement>, val?: string) => {
      const placeholderChar = '_';
      if (val && !val.includes(placeholderChar)) {
        const d = translateToDate(inputFormat, val);
        if (d) setDate(d);
      }
    }

    const onClearHandler = () => {
      setDate(undefined);
    }

    const onDateChangeHandler = (d?: Date) => {
      if (d) setDate(d);
      if (onDateChange) onDateChange(date)
    }

    const trigger = (
      <InputMask
        name={"datepicker"}
        mask={{ name: 'date', type: inputFormat }}
        validator={{ name: 'date', type: inputFormat }}
        placeholder={inputFormat}
        value={date ? translateToString(inputFormat, date) : ''}
        onChange={onChangeHandler}
        onClear={onClearHandler}
      />
    );

    return (
      <Popover
        trigger={trigger}
        position='bottom-start'
        on='click'
        appendToBody={true}
      >
        <Calendar
          {...rest}
          date={date}
          onDateChange={onDateChangeHandler}
        />
      </Popover>
    );
  } else {
    return (
      <Calendar
        {...rest}
        date={date}
        onDateChange={onDateChange}
      />
    );
  }
};

export default DatePicker;
