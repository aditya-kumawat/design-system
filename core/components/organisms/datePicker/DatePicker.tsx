import * as React from 'react';
import { Calendar, SharedProps } from '../calendar/Calendar';
import { DateType, DateFormat } from '../calendar/types';
import Popover, { Position } from '@/components/molecules/popover';
import { InputMask, InputMaskProps } from '@/components/molecules/inputMask';
import { Utils } from '@/index';
import { Validators } from '@/utils/types';
import { convertToDate, translateToDate, translateToString, compareDate, getDateInfo } from '../calendar/utility';

export type DatePickerProps = SharedProps & {
  /**
   * Callback function called when date is changed
   * @argument date Date object
   * @argument dateVal Date string value as per `outputFormat`
   */
  onDateChange?: (date: Date | undefined, dateVal?: string) => void;
  /**
   * Selected date
   *
   * `number` - number of milliseconds elapsed since January 1, 1970, 00:00:00 UTC
   *
   * `string` - Date string value as per `inputFormat`
   */
  date?: DateType;
  /**
   * Set if `InputMask` should be used as trigger
   */
  withInput?: boolean;
  /**
   * Sets open state of `Popover`
   * @default false
   */
  open?: boolean;
  /**
   * Position of `DatePicker` with respect to `InputMask`
   */
  position: Position;
  /**
   * Should be used if `date` is of type `string`
   */
  inputFormat: DateFormat;
  /**
   * Should be used to translate `date` to desired format for `onDateChange` callback
   */
  outputFormat: DateFormat;
  /**
   * Props to be used for `InputMask`
   */
  inputOptions: Omit<InputMaskProps, 'mask' | 'value' | 'onChange' | 'onBlur' | 'onClear' | 'error'>;
  /**
   * custom Validator for `DatePicker`
   *
   * `ValidatorFn = (val: string, format: string) => boolean`
   */
  validators: Validators;
  /**
   * Close Popover on date selection
   */
  closeOnSelect: boolean;
};

interface DatePickerState {
  init: boolean;
  date?: Date;
  error: boolean;
  open: boolean;
}

export class DatePicker extends React.Component<DatePickerProps, DatePickerState> {
  static defaultProps = {
    ...Calendar.defaultProps,
    position: 'bottom-start',
    inputFormat: 'mm/dd/yyyy',
    outputFormat: 'mm/dd/yyyy',
    validators: [Utils.validators.date],
    inputOptions: {},
    closeOnSelect: true
  };

  constructor(props: DatePickerProps) {
    super(props);

    const {
      inputFormat,
      validators
    } = props;

    const date = convertToDate(props.date, inputFormat, validators);
    const error = this.getError(date);

    this.state = {
      date,
      error,
      init: false,
      open: props.open || false,
    };
  }

  componentDidUpdate(prevProps: DatePickerProps, prevState: DatePickerState) {
    if (prevProps.date !== this.props.date) {
      const {
        inputFormat,
        validators
      } = this.props;

      const d = convertToDate(this.props.date, inputFormat, validators);
      this.setState({
        date: d
      });
    }

    if (prevProps.open !== this.props.open) {
      this.setState({
        open: this.props.open || false
      });
    }

    if (prevState.date !== this.state.date) {
      const {
        onDateChange,
        outputFormat
      } = this.props;

      const {
        date
      } = this.state;

      const newError = this.getError(date);

      this.setState({ error: newError });

      if (onDateChange) {
        if (!newError) {
          const dVal = translateToString(outputFormat, date);
          onDateChange(date, dVal);
        } else {
          onDateChange(undefined, '');
        }
      }
    }
  }

  getError = (date?: Date) => {
    const {
      disabledBefore,
      disabledAfter
    } = this.props;

    const {
      year: dbYear,
      month: dbMonth,
      date: dbDate
    } = getDateInfo(disabledBefore);

    const {
      year: daYear,
      month: daMonth,
      date: daDate
    } = getDateInfo(disabledAfter);

    return !date ? true
      : compareDate(date, 'less', dbYear, dbMonth, dbDate)
      || compareDate(date, 'more', daYear, daMonth, daDate);
  }

  onDateChangeHandler = (d?: Date) => {
    this.setState({
      init: true,
      date: d
    });

    const {
      closeOnSelect
    } = this.props;

    if (closeOnSelect) this.setState({ open: false });
  }

  onChangeHandler = (_e: React.ChangeEvent<HTMLInputElement>, val?: string) => {
    const {
      inputFormat,
      validators,
      inputOptions
    } = this.props;

    this.setState({
      open: true
    });

    const { placeholderChar = '_' } = inputOptions;
    if (val && !val.includes(placeholderChar)) {
      const d = translateToDate(inputFormat, val, validators);
      this.setState({ date: d });
    }
  }

  onFocusHandler = () => {
    this.setState({
      init: true
    });
  }

  onBlurHandler = (_e: React.ChangeEvent<HTMLInputElement>, val?: string) => {
    const { placeholderChar = '_' } = this.props.inputOptions;
    if (!val || val.includes(placeholderChar)) {
      this.setState({ date: undefined });
    }
  }

  onClearHandler = () => {
    this.setState({
      init: true,
      date: undefined
    });
  }

  onToggleHandler = (o: boolean, type?: string) => {
    switch (type) {
      case 'outsideClick':
        this.setState({ open: o });
        break;
      case 'onClick':
        this.setState({ open: true });
        break;
    }
  }

  renderCalendar() {
    const {
      date: dateProp,
      open,
      position,
      inputFormat,
      outputFormat,
      inputOptions,
      validators,
      withInput,
      disabledBefore,
      disabledAfter,
      onDateChange,
      closeOnSelect,
      ...rest
    } = this.props;

    const {
      date
    } = this.state;

    return (
      <Calendar
        {...rest}
        date={convertToDate(date, inputFormat, validators)}
        disabledBefore={convertToDate(disabledBefore, inputFormat, validators)}
        disabledAfter={convertToDate(disabledAfter, inputFormat, validators)}
        onDateChange={this.onDateChangeHandler}
      />
    );
  }

  render() {
    const {
      position,
      inputFormat,
      inputOptions,
      withInput,
      validators
    } = this.props;

    const {
      init,
      date,
      error,
      open
    } = this.state;

    if (withInput) {
      const showError = inputOptions.required && error && init;
      const inputValidator = (val: string): boolean => {
        return Utils.validators.isValid(validators, val, inputFormat);
      };

      const trigger = (
        <InputMask
          icon="events"
          placeholder={inputFormat}
          {...inputOptions}
          error={showError}
          mask={Utils.masks.date[inputFormat]}
          value={date ? translateToString(inputFormat, date) : ''}
          onChange={this.onChangeHandler}
          onFocus={this.onFocusHandler}
          onBlur={this.onBlurHandler}
          onClear={this.onClearHandler}
          caption={showError ? inputOptions.caption || 'Invalid value' : ''}
          validators={[inputValidator]}
        />
      );

      return (
        <Popover
          trigger={trigger}
          triggerClass="w-100"
          position={position}
          appendToBody={true}
          open={open}
          onToggle={this.onToggleHandler}
        >
          {this.renderCalendar()}
        </Popover>
      );
    }

    return this.renderCalendar();
  }
}

export default DatePicker;
