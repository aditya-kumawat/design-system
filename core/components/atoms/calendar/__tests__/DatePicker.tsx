import * as React from 'react';
import { shallow } from 'enzyme';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';
import DatePicker, { IDatePickerProps as IProps } from '../DatePicker';

const view = ['year', 'month', 'date'];
const booleanValue = [true, false];

describe('DatePicker component', () => {
  const mapper: Record<string, any> = {
    view: valueHelper(view, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as IProps;

    it(testMessageHelper(attr), () => {
      const tree = shallow(
        <DatePicker
          date={new Date(2020, 2, 1)}
          {...attr}
        >
          DatePicker
        </DatePicker >
      );
      expect(tree).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('DatePicker component', () => {
  const mapper: Record<string, any> = {
    withInput: valueHelper(booleanValue, { required: true, iterate: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as IProps;

    it(testMessageHelper(attr), () => {
      const tree = shallow(
        <DatePicker
          date={new Date(2020, 2, 1)}
          {...attr}
        >
          DatePicker
        </DatePicker >
      );
      expect(tree).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});
