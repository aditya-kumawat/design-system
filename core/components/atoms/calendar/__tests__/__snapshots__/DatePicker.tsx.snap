// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker component 
  view: "date"
 1`] = `
<Component
  onDateChange={[Function]}
  view="date"
>
  DatePicker
</Component>
`;

exports[`DatePicker component 
  view: "month"
 1`] = `
<Component
  onDateChange={[Function]}
  view="month"
>
  DatePicker
</Component>
`;

exports[`DatePicker component 
  view: "year"
 1`] = `
<Component
  onDateChange={[Function]}
  view="year"
>
  DatePicker
</Component>
`;

exports[`DatePicker component 
  withInput: false
 1`] = `
<Component
  onDateChange={[Function]}
>
  DatePicker
</Component>
`;

exports[`DatePicker component 
  withInput: true
 1`] = `
<Popover
  appendToBody={true}
  onToggle={[Function]}
  open={false}
  trigger={
    <ForwardRef
      error={false}
      mask={
        Array [
          /\\\\d/,
          /\\\\d/,
          "/",
          /\\\\d/,
          /\\\\d/,
          "/",
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
          /\\\\d/,
        ]
      }
      name="datepicker"
      onBlur={[Function]}
      onChange={[Function]}
      onClear={[Function]}
      placeholder="dd/mm/yy"
      placeholderChar="_"
      value=""
    />
  }
>
  <Component
    onDateChange={[Function]}
  >
    DatePicker
  </Component>
</Popover>
`;
